### DISCUSSION TITLE
Dialogs

### DISCUSSION THREAD

**Discussion by JordanMarr** [JordanMarr/ReactiveElmish.Avalonia#46] [General]
Making a note here to try out this library for creating embedded dialogs that looks very nice:

https://github.com/AvaloniaUtils/DialogHost.Avalonia

**Comment by houstonhaynes** [comment]
I think that's a great idea. Have you looked at [FluentAvalonia](https://github.com/amwx/FluentAvalonia)? They have a range of dialogs - of course I'm already nibbling around the edges with it - so I'm a bit "vested" in it already. 😸 
![Screenshot 2024-01-06 at 1 35 00 PM](https://github.com/JordanMarr/ReactiveElmish.Avalonia/assets/8174976/1dad33ec-fe9f-4112-a610-22bd07842efe)
![Screenshot 2024-01-06 at 1 35 41 PM](https://github.com/JordanMarr/ReactiveElmish.Avalonia/assets/8174976/97fd5f66-076f-4e8b-90a6-b5cb1cd728d0)

**Reply by houstonhaynes** [reply]
```xml
<!-- TaskDialog can be declared in Xaml -->
<!-- However, it must still be launched in C# -->
<ui:TaskDialog Name="TaskDialog1"
                Title="FluentAvalonia"
                Header="Header Here"
                SubHeader="SubHeader"
                Content="This is some sample text, but you can put more advanced content here if you like"
                FooterVisibility="Always"
                IsFooterExpanded="False"
                ShowProgressBar="False">
    <ui:TaskDialog.Commands>
        <ui:TaskDialogCommand Text="Sample Command"
                                Description="This is a description of the Sample Command"
                                IsEnabled="True"
                                DialogResult="CommandResult"/>
    </ui:TaskDialog.Commands>

    <ui:TaskDialog.Buttons>
        <ui:TaskDialogButton Text="OK" DialogResult="OK" />
        <ui:TaskDialogButton Text="Cancel" DialogResult="Cancel" />
    </ui:TaskDialog.Buttons>

    <ui:TaskDialog.Footer>
        <CheckBox Content="Never show me this again" />
    </ui:TaskDialog.Footer>
</ui:TaskDialog>
```

```csharp
// If your TaskDialog is declared in Xaml, just reference it and call ShowAsync()
// TaskDialog can launch in either windowed or hosted mode. Windowed mode presents the dialog in a
// top level window (using CoreWindow), thus it has a titlebar and an additional property Title
// Hosted mode uses the OverlayLayer.

this.FindControl<TaskDialog>("TaskDialog1").ShowAsync();

// Declaring a TaskDialog from C#:
var td = new TaskDialog
{
    // Title property only applies on Windowed dialogs
    Title = "FluentAvalonia",
    Header = "Header",
    Subheader = "Subheader",
    Content = "This is some sample text, but you can put more advanced content here if you like",
    IconSource = new SymbolIconSource { Symbol = Symbol.Save },
    FooterVisibility = TaskDialogFooterVisibility.Auto,
    Footer = new CheckBox { Content = "Never show me this again" },
    Commands =
    {
        new TaskDialogCommand
        {
            Text = "Command Text",
            Description = "Description",
            DialogResult = "CommandResult",
            // ClosesOnInvoked property lets you choose if invoking this command closes the dialog
            // automatically (default true)
            ClosesOnInvoked = true,
            // Can also set IconSource
            // Can also set IsEnabled
        }
    },
    Buttons =
    {
        // For more advanced scenarios, you can create your own buttons
        // Custom buttons allow you to attach icons, custom results,
        // command and click handlers to fully customize your experience
        // Note: 'null' is not a valid dialog result and is automatically
        // converted to TaskDialogStandardResult.None when the dialog closes
        new TaskDialogButton("OK" /* text */, "myResult" /* dialogResult */)


        // There are some default buttons for simple cases provided
        // These have predefinded text and results that correspond to
        // TaskDialogStandardResult enum
        // Note that built in buttons cannot have Commands, Icons, or
        // click handlers attached to them
        TaskDialogButton.OKButton,
        TaskDialogButton.CancelButton,
        TaskDialogButton.YesButton,
        TaskDialogButton.NoButton,
        TaskDialogButton.RetryButton,
        TaskDialogButton.CloseButton
    }
};

// Before showing a dialog declared in C#, you MUST set the XamlRoot property
// Using the VisualRoot is fine, if the VisualRoot is a Window, the dialog automatically launches in
// Windowed mode, otherwise, it tries to find the OverlayLayer and will launch in hosted mode
// If your TaskDialog is declared in Xaml, this is automatically handled for you
td.XamlRoot = this.VisualRoot;

var result = await td.ShowAsync();

// If you want to force hosted mode, ShowAsync accepts a parameter 'showHosted' to force this mode
var result = await td.ShowAsync(true);
```

**Comment by daz10000** [comment]
Nice!  I have been pondering best way to do those light weight interactions. I don’t want to build a whole new variety for a simple interaction.  The modal widget in silverlight was really helpful back in the day but hadn’t seen an avalonia equivalent. Am 1/6/24 um 10:37 AM schrieb Houston Haynes ***@***.***>:﻿


<ui:TaskDialog Name="TaskDialog1"
                Title="FluentAvalonia"
                Header="Header Here"
                SubHeader="SubHeader"
                Content="This is some sample text, but you can put more advanced content here if you like"
                FooterVisibility="Always"
                IsFooterExpanded="False"
                ShowProgressBar="False">
    <ui:TaskDialog.Commands>
        <ui:TaskDialogCommand Text="Sample Command"
                                Description="This is a description of the Sample Command"
                                IsEnabled="True"
                                DialogResult="CommandResult"/>
    </ui:TaskDialog.Commands>

    <ui:TaskDialog.Buttons>
        <ui:TaskDialogButton Text="OK" DialogResult="OK" />
        <ui:TaskDialogButton Text="Cancel" DialogResult="Cancel" />
    </ui:TaskDialog.Buttons>

    <ui:TaskDialog.Footer>
        <CheckBox Content="Never show me this again" />
    </ui:TaskDialog.Footer>
</ui:TaskDialog>
// If your TaskDialog is declared in Xaml, just reference it and call ShowAsync()
// TaskDialog can launch in either windowed or hosted mode. Windowed mode presents the dialog in a
// top level window (using CoreWindow), thus it has a titlebar and an additional property Title
// Hosted mode uses the OverlayLayer.

this.FindControl<TaskDialog>("TaskDialog1").ShowAsync();

// Declaring a TaskDialog from C#:
var td = new TaskDialog
{
    // Title property only applies on Windowed dialogs
    Title = "FluentAvalonia",
    Header = "Header",
    Subheader = "Subheader",
    Content = "This is some sample text, but you can put more advanced content here if you like",
    IconSource = new SymbolIconSource { Symbol = Symbol.Save },
    FooterVisibility = TaskDialogFooterVisibility.Auto,
    Footer = new CheckBox { Content = "Never show me this again" },
    Commands =
    {
        new TaskDialogCommand
        {
            Text = "Command Text",
            Description = "Description",
            DialogResult = "CommandResult",
            // ClosesOnInvoked property lets you choose if invoking this command closes the dialog
            // automatically (default true)
            ClosesOnInvoked = true,
            // Can also set IconSource
            // Can also set IsEnabled
        }
    },
    Buttons =
    {
        // For more advanced scenarios, you can create your own buttons
        // Custom buttons allow you to attach icons, custom results,
        // command and click handlers to fully customize your experience
        // Note: 'null' is not a valid dialog result and is automatically
        // converted to TaskDialogStandardResult.None when the dialog closes
        new TaskDialogButton("OK" /* text */, "myResult" /* dialogResult */)


        // There are some default buttons for simple cases provided
        // These have predefinded text and results that correspond to
        // TaskDialogStandardResult enum
        // Note that built in buttons cannot have Commands, Icons, or
        // click handlers attached to them
        TaskDialogButton.OKButton,
        TaskDialogButton.CancelButton,
        TaskDialogButton.YesButton,
        TaskDialogButton.NoButton,
        TaskDialogButton.RetryButton,
        TaskDialogButton.CloseButton
    }
};

// Before showing a dialog declared in C#, you MUST set the XamlRoot property
// Using the VisualRoot is fine, if the VisualRoot is a Window, the dialog automatically launches in
// Windowed mode, otherwise, it tries to find the OverlayLayer and will launch in hosted mode
// If your TaskDialog is declared in Xaml, this is automatically handled for you
td.XamlRoot = this.VisualRoot;

var result = await td.ShowAsync();

// If you want to force hosted mode, ShowAsync accepts a parameter 'showHosted' to force this mode
var result = await td.ShowAsync(true);

**Comment by JordanMarr** [comment]
Those look great. Considering how nice the hamburger navigation menu is, it makes sense to also utilize the Fluent Avalonia dialogs.

**Reply by houstonhaynes** [reply]
@Jamie_Rower and I have gotten a few "paper cuts" around implementing FluentAvalonia and my guess is we'll have to revisit some of them as we smooth out the rough edges. That said - adding the NavigationView and placing that "inside" the "content" portion of the Avalonia _native_ SplitView was pretty straightforward. 

One of the places I was considering using the Task Dialog was popping up a BIG modal window from button embedded a response from the "Aiden" agent - where the finding could be viewed as a chart similar to a user-driven in-app visual - but simply pops over the main UI for a live "at a glance" view. Looking at those controls makes all of that attainable without contemplating too much blood sweat and tears.

It's worthwhile to spend a bit of quality time with the FA sample app - there were some things that we *definitely* want to implement. The other option was something we were dreading - having to go through a bunch of one-off libraries that are showing their age (AuraUI) and try to piece things together. There's a half-dozen things in there that are right-at-hand for behaviors/features we want to implement.

**Reply by houstonhaynes** [reply]
The "flip side" of the sample app is that some of the "code" that they show with the samples are woefully incomplete (looking at you, FontIconSource definitions)

**Reply by houstonhaynes** [reply]
https://youtu.be/NnF6YRN4B5o?si=IrKBzWPb3d_N55Fl
